### ISSUE TITLE
attached type parameters

### ISSUE THREAD

**Issue by aep** [zetzit/zz#121]
generic containers work, but i still hate them somehow. i thought not knowing the interior type is fine because thats what golang does too, but then golang has reflection so you can get the interior at runtime.

this is unfixably unsafe

```C++
    new+10  v = map::make();

    u8 val =12;
    v.insert("hello", &val);

    u64 * bad = v.get("hello"); 

```
typestate could make it safe

```C++
    new+10  v = map::make();
    static_attest(map::value(&v, u8));

    u8 val =12;
    v.insert("hello", &val);

    u64 * bad = v.get("hello"); // error: u64 != u8
```

But they become annoying quickly, because typestate needs to be manually carried around everywhere you use the exterior type.

Other languages do "templates" so they just create a new type for each possible parameter. That's horribly inefficient, as each instance also has to have all of its functions duplicated. It also doesnt map to C.

Instead i'm proposing to just attach the interiour typestate to the external typestate permanently somehow, creating a new virtual type. That's essentially just what tail already is, a permanent type state.

One possible syntax is just reusing the familiar template angled brackets. however, the tail market has to go somewhere

```C++
struct Vec<type Item>

```

**Comment by aep** [comment]
@wrl

**Comment by jwerle** [comment]
Knowing the interior type for things like `Vec` is definitely important. Is `typeid` in theory syntax supported at the moment?

I like the usage of '[]' instead of '<>' because of the template syntax trap you mentioned. Readers, new comers, and possible new contributors will need this called out.

Permanent theories look interesting, but hard to follow based on the syntax in the example above.

**Comment by aep** [comment]
i'm stuck figuring out how attached parameters and smt expressions mix.

this is fine

```C++
export fn push(Vec<A> *self)
   where A !=  typeof(int)
```

but you cannot deconstruct a nested type

```C++
export fn push(Vec<Map<String, A>> *self)
   where A != typeof(int)
```

because 
  1. A is ambiguously either a type or a bind
  2. there's no way to specify another push(Vec<anything not map>). you can only have one function with that name

so the only sure way is interior type access in smt expressions

```C++
export fn push(Vec *self)
   where map::value(vec::item(self)) !=  typeof(int)
```

that solves 1, but it does not solve 2
and maybe you actually wanted to use the type at runtime for something

```C++
export fn push(Vec<Map<String, A>> *self)
{
    let more_mem = sizeof(A);
}
```



so the best way i can come up with right now is make callsite generated arguments explicitly available to user code (the'res currently only the builtin callsite<line>

```C++
export fn push(Vec *self, typeid X)
   where X = map::value(vec::item(self))
{
    let more_mem = sizeof(A);
}
```

note the assign = instead of compare == indicating this is assigned from evaluation in the callsite 

callsite generation is something you normally never use, so its ok to require some more advanced understanding of the smt call order. For daily usage, we can shortcut the permanent typestate expression

```C++

struct Vec {
}
theory item(Vec self) typeid;


pub fn main() {

   vec::Vec<vec::item = int, +100> v;

   new <vec::item = int,+100> v = vec::make()
   
   static_asset(vec::item(v) == int);
}
```

at least i hope i can make this work. it requires being able to skip typeof().
the expanded expression looks like this:

```C++
   Vec<vec::item(self) == typeof(int), tail(self) == 100> v;
```

this is a full smt expression, so we basically have touring complete types. It's mind bending actually.


it _might_ be possible to reduce the shortcut further by assuming unqualified names inside an angled bracket refer to the same module as the exteriour.

```C++
struct A {
   Vec<item = int, +100> v;
}
```

**Comment by jwerle** [comment]
Wow! The power behind this would be powerful, if possible. I like the syntax
